beluga/include/beluga/estimation/cluster_based_estimator.hpp

```diff
@@ -0,0 +1,372 @@
+  }
+
+  return cluster_estimates;
+}
+
+}  // namespace cse_detail
+
+/// Primary template for a cluster-based estimation algorithm.
+/**
+ * Particles are groups into clusters around local maximums and the mean and covariance of the cluster with the highest
+ * weight is returned.
+ *
+ * This class implements the EstimationInterface interface
+ * and satisfies \ref StateEstimatorPage.
+ * *
+ * \tparam Mixin The mixin that implements the particle filter interface.
+ * \tparam StateType The type of the state to be estimated.
+ * \tparam CovarianceType The type of the covariance matrix of the state to be estimated.
+ */
+template <class Mixin, class StateType, class CovarianceType>
+class ClusterBasedStateEstimator : public Mixin {
+ public:
+  /// Constructs a ClusterBasedStateEstimator instance.
+  /**
+   * \tparam ...Args Arguments types for the remaining mixin constructors.
+   * \param ...args Arguments that are not used by this part of the mixin, but by others.
+   */
+  template <class... Args>
+  explicit ClusterBasedStateEstimator(Args&&... args) : Mixin(std::forward<Args>(args)...) {}
+
+  /// \copydoc EstimationInterface2d::estimate()
+  [[nodiscard]] std::pair<StateType, CovarianceType> estimate() const final;
+
+ private:
+  static constexpr double kSpatialHashResolution = 0.20;   ///< clustering algorithm spatial resolution
+  static constexpr double kAngularHashResolution = 0.524;  ///< clustering algorithm angular resolution
+
+  /// @brief spatial hash function used to group particles in cells
+  const beluga::spatial_hash<Sophus::SE2d> spatial_hash_function_{
+      kSpatialHashResolution,  // x
+      kSpatialHashResolution,  // y
+      kAngularHashResolution   // theta
+  };
+
+  /// @brief Adjacent grid cells used by the clustering algorithm.
+  const std::vector<Sophus::SE2d> adjacent_grid_cells_ = {
+      Sophus::SE2d{Sophus::SO2d{0.0}, Sophus::Vector2d{+kSpatialHashResolution, 0.0}},
+      Sophus::SE2d{Sophus::SO2d{0.0}, Sophus::Vector2d{-kSpatialHashResolution, 0.0}},
+      Sophus::SE2d{Sophus::SO2d{0.0}, Sophus::Vector2d{0.0, +kSpatialHashResolution}},
+      Sophus::SE2d{Sophus::SO2d{0.0}, Sophus::Vector2d{0.0, -kSpatialHashResolution}},
+      Sophus::SE2d{Sophus::SO2d{+kAngularHashResolution}, Sophus::Vector2d{0.0, 0.0}},
+      Sophus::SE2d{Sophus::SO2d{-kAngularHashResolution}, Sophus::Vector2d{0.0, 0.0}},
+  };
+};
+
+template <class Mixin, class StateType, class CovarianceType>
+std::pair<StateType, CovarianceType> ClusterBasedStateEstimator<Mixin, StateType, CovarianceType>::estimate() const {
+  auto hashes = cse_detail::precalculate_particle_hashes(this->self().states(), spatial_hash_function_);
+  auto grid_cell_data = cse_detail::populate_grid_cell_data_from_particles<cse_detail::GridCellDataMap2D>(
+      this->self().states(), this->self().weights(), hashes);
+  const auto weight_cap = cse_detail::calculate_percentile_weight_threshold(grid_cell_data, 0.9);
```

Comment:
`0.9` should probably be a documented parameter too.

Reply:
Agreed.